### PR TITLE
Split NICO++ according to the paper 'Change is Hard'

### PR DESCRIPTION
**Pull request by tmke8**

### CODE REVIEW

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 1]
```diff
@@ -110,6 +110,8 @@ class NICOPP(CdtVisionDataset[TernarySample, Tensor, Tensor]):
     SampleType: TypeAlias = TernarySample
     Superclass: TypeAlias = NicoPPTarget
     Subclass: TypeAlias = NicoPPAttr
+    data_split_seed: ClassVar[int] = 666  # this is the seed from the paper
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:113]: Using the same seed won't lead to consistent splits but I guess 666 is as sensible a default as any, despite the numerological associations

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 2]
```diff
@@ -188,7 +190,7 @@ def _extract_metadata(self):
         meta = json.load(open(self._base_dir / "dg_label_id_mapping.json", "r"))
 
         def _make_balanced_testset(
-            df: pd.DataFrame, seed: int = 666, verbose: bool = True, num_samples_val_test: int = 75
+            df: pd.DataFrame, seed: int, num_samples_val_test: int, verbose: bool = True
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:193]:
```suggestion
            df: pd.DataFrame, *, seed: int, num_samples_val_test: int, verbose: bool = True
```

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 3]
```diff
@@ -152,90 +139,102 @@ def __init__(
...
-            columns={0: "subclass", 1: "superclass", 2: "filename"}
+        if not all(
+            (self._base_dir / "public_dg_0416" / "train" / attr).exists() for attr in NicoPPAttr
+        ):
+            return False
+        if not (self._base_dir / "dg_label_id_mapping.json").exists():
+            return False
+        return True
+
+    def _extract_metadata(self):
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:187]:
```suggestion
    def _extract_metadata(self) -> None:
```

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 4]
```diff
@@ -152,90 +139,102 @@ def __init__(
...
+        attributes = ["autumn", "dim", "grass", "outdoor", "rock", "water"]  # 6 attrs, 60 labels
+        meta = json.load(open(self._base_dir / "dg_label_id_mapping.json", "r"))
+
+        def _make_balanced_testset(
+            df: pd.DataFrame, seed: int, num_samples_val_test: int, verbose: bool = True
+        ) -> pd.DataFrame:
+            # each group has a test set size of (2/3 * num_samples_val_test) and a val set size of
+            # (1/3 * num_samples_val_test); if total samples in original group < num_samples_val_test,
+            # val/test will still be split by 1:2, but no training samples remained
+            import random
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:198]: Does this import need to be local?

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 5]
```diff
@@ -152,90 +139,102 @@ def __init__(
...
+            random.seed(seed)
+            val_set, test_set = [], []
+            for g in pd.unique(df["g"]):
+                df_group = df[df["g"] == g]
+                curr_data = df_group["filename"].values
+                random.shuffle(curr_data)
+                split_size = min(len(curr_data), num_samples_val_test)
+                val_set += list(curr_data[: split_size // 3])
+                test_set += list(curr_data[split_size // 3 : split_size])
+            if verbose:
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:209]:
```suggestion
            if verbose:
```
Could probably do away with this verbose mode (or at least the toggleability of it)

File: conduit/data/datasets/vision/nico_plus_plus.py (hunk 4 above)
Comment (tmke8) [reply conduit/data/datasets/vision/nico_plus_plus.py:198]: I guess the only reason is that it's rarely needed.

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 6]
```diff
@@ -152,90 +139,102 @@ def __init__(
...
+            combined_set.update(dict(zip(test_set, [2 for _ in range(len(test_set))])))
+            df["split"] = df["filename"].map(combined_set)
+            df["split"].fillna(0, inplace=True)
+            df["split"] = df.split.astype(int)
+            return df
+
+        all_data = []
+        for c, attr in enumerate(attributes):
+            for label in meta:
+                folder_path = self._base_dir / "public_dg_0416" / "train" / attr / label
```
Comment (MylesBartlett) [inline conduit/data/datasets/vision/nico_plus_plus.py:222]: Maybe make "public_dg_0416" a class variable?

Comment (tmke8) [reply conduit/data/datasets/vision/nico_plus_plus.py:222]: something like

```python
    subpath: ClassVar[Path] = Path("public_dg_0416") / "train"
```
?

Comment (MylesBartlett) [reply conduit/data/datasets/vision/nico_plus_plus.py:222]: Sure, if only the full subpath is ever needed (which seems to be the case)

**Review by MylesBartlett — APPROVED** [review]

File: conduit/data/datasets/vision/nico_plus_plus.py (hunk 1 above)
Comment (tmke8) [reply conduit/data/datasets/vision/nico_plus_plus.py:114]: who would have thought that we're implementing a method from Satanists

File: conduit/data/datasets/vision/nico_plus_plus.py [hunk 7]
```diff
@@ -152,90 +143,100 @@ def __init__(
...
+                curr_data = df_group["filename"].values
+                random.shuffle(curr_data)
+                split_size = min(len(curr_data), num_samples_val_test)
+                val_set += list(curr_data[: split_size // 3])
+                test_set += list(curr_data[split_size // 3 : split_size])
+            self.logger.info(f"Val: {len(val_set)}, Test: {len(test_set)}")
+            assert len(set(val_set).intersection(set(test_set))) == 0
+            combined_set = dict(zip(val_set, [NicoPPSplit.VAL.value for _ in range(len(val_set))]))
+            combined_set.update(
+                dict(zip(test_set, [NicoPPSplit.TEST.value for _ in range(len(test_set))]))
```
Comment (tmke8) [inline conduit/data/datasets/vision/nico_plus_plus.py:214]: If I understand this code correctly, it is incredibly inefficient, but whatever.

**Pull request merged**